### PR TITLE
Various fixes in java.net.URI

### PR DESCRIPTION
**Pull request by exoego**
Noticed these when using `URI` and then running tests from [Apache Harmony](https://github.com/apache/harmony) which is licensed under Apache License 2.0.

Changes
1. IPv4 should not accept the digits greater than 255 (e.g. `256.384.512.777`)
2.  Normalize host in URI.hashCode.
    * This is necessary so that `foo.com` and `FoO.CoM` (which compare equal) produce the same hashCode.
3. Properly compare URIs.
    1. Apply case insensitivity when comparing URI parts
    2. Compare paths, query and fragments when comparing URI parts. This is necessary so that `foo.com/foo` and `FoO.CoM/foo` equals, but `foo.com/foo` and `FoO.CoM/FoO` does not.
    3. Return 0 if both scheme is undefined

### CODE REVIEW

File: javalib/src/main/scala/java/net/URI.scala [hunk 1]
```diff
@@ -107,16 +107,22 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
+    if (this._scheme != that._scheme) {
+      val s = this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
+      if (s == 0 && isOpaque()) cmpSsp()
+      else s
+    } else if (this._isOpaque != that._isOpaque)
+      // A hierarchical URI is less than an opaque URI
+      if (this._isOpaque) 1 else -1
+    else if (_isOpaque)
+      cmpSsp()
+    else if (this._authority != that._authority) {
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: While you are here, please put braces around all if / else blocks (unfortunately this seems undocumented but is a style we follow).

File: javalib/src/main/scala/java/net/URI.scala [hunk 2]
```diff
@@ -107,16 +107,22 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
+    @inline def cmpSsp(): Int =  {
       val ssp = cmp(this._schemeSpecificPart, that._schemeSpecificPart)
       if (ssp != 0) ssp
       else cmpOpt(this._fragment, that._fragment)
-    } else if (this._authority != that._authority) {
+    }
+
+    if (this._scheme != that._scheme) {
+      val s = this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
+      if (s == 0 && isOpaque()) cmpSsp()
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: I'm not sure this now works properly if two URIs have the same scheme but one is opaque an the other is not.

Consider replacing all of this with something like this:

```scala
List(
  this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase)),
  this._isOpaque.toInt - that._isOpaque.toInt, // A hierarchical URI is less than an opaque URI
  ...
).find(_ != 0).getOrElse(0)
```

This makes it unmistakably clear that these are merely priorities for comparison. Also, it will fix similar bugs in other parts when the supplied `cmp` is used.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala [hunk 3]
```diff
@@ -341,6 +341,10 @@ class URITest {
       assertEquals(new URI("http://example.com/asdf%6a").hashCode,
           new URI("http://example.com/asdf%6A").hashCode)
     }
+    assertEquals(new URI("http://example.com").hashCode(),
+      new URI("http://Example.CoM").hashCode())
+    assertNotEquals(new URI("http://example.com/example-com").hashCode(),
+      new URI("http://Example.CoM/eXAMplE-cOm").hashCode())
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: It seems these tests need some maintenance:

- `executingInJVM` should be dropped (we dropped JDK6/7 support).
- Tests for `equals` and `hashCode` really should use the same test cases since the specification is so intimately related.

I'd be great if you could add another commit for this.

File: javalib/src/main/scala/java/net/URI.scala [hunk 4]
```diff
@@ -113,6 +113,15 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
 
+    @inline def cmpPathQueryFragment(): Int = {
+      if (this._path != that._path)
+        cmpOpt(this._path, that._path)
+      else if (this._query != that._query)
+        cmpOpt(this._query, that._query)
+      else
+        cmpOpt(this._fragment, that._fragment)
+    }
+
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: Same comment than on the previous commit. Probably easier to fix this in a more principled fashion.

File: javalib/src/main/scala/java/net/URI.scala (hunk 1 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Fixed

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala (hunk 3 above)
Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: Dropped `executingInJVM` in https://github.com/scala-js/scala-js/pull/3690/commits/065ef673dd763d427160d3365b1eaf4f76f55adb

Refactoring tests for `equals` and `hashCode` in https://github.com/scala-js/scala-js/pull/3690/commits/b9717e97663d6d88eb9184128be7d96e41cf4d01

File: javalib/src/main/scala/java/net/URI.scala (hunk 2 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Refactored in https://github.com/scala-js/scala-js/pull/3690/commits/b2caa31c242ad85efee3fe748e9a7b925f220a0c and fixed an existing issue in subequent https://github.com/scala-js/scala-js/pull/3690/commits/52d33ceece6e3fb4b655a1ccc0dbf7c4df2fe7d7

File: javalib/src/main/scala/java/net/URI.scala (hunk 4 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Refactored this same as above in https://github.com/scala-js/scala-js/commit/b2caa31c242ad85efee3fe748e9a7b925f220a0c

File: javalib/src/main/scala/java/net/URI.scala [hunk 5]
```diff
@@ -106,36 +106,45 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
-      } else
-        cmpOpt(this._authority, that._authority)
-    } else if (this._path != that._path)
-      cmpOpt(this._path, that._path)
-    else if (this._query != that._query)
-      cmpOpt(this._query, that._query)
-    else
-      cmpOpt(this._fragment, that._fragment)
+      } else cmpOpt(this._authority, that._authority)
+    ).find(_ != 0).getOrElse(0)
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: I think the current version is kind of hard to read.

How about something like this:

```scala
List(
    cmpOpt(this._scheme, that._scheme)(_.compareToIgnoreCase),
    this.isOpaque.compareTo(that.isOpaque), // A hierarchical URI is less than an opaque URI
) ++ (
    if (this.isOpaque) {
      List(
        cmp(this._schemeSpecificPart, that._schemeSpecificPart),
      )
    } else if (this._host.isDefined && that._host.isDefined) {
      List(
        cmpOpt(this._userInfo, that._userInfo)(cmp),
        cmpOpt(this._host, that._host)(_.compareToIgnoreCase),
        cmpOpt(this._port, that._port)(_ - _),  // change _port to js.UndefOr[Int] and fold into -1 only when getPort is called.
      )
    } else {
      List(
         cmpOpt(this._authority, that._authority)(cmp)
      )
    }
) ++ List(
    cmpOpt(this._path, that._path)(cmp),
    cmpOpt(this._query, that._query)(cmp),
    cmpOpt(this._fragment, that._fragment)(cmp),
).find(_ != 0).getOrElse(0)
```

It is unfortunately not as nice as I anticipated, mostly because of the if-branches. Not sure how to improve though. Maybe something with `Option#orElse` or helper functions would work better.

Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Refactoring so and cleanup commits.

File: javalib/src/main/scala/java/net/URI.scala [hunk 6]
```diff
@@ -101,41 +101,37 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
+    ) ++ (
+      if (this.isOpaque()) {
+        List(
+          cmp(this._schemeSpecificPart, that._schemeSpecificPart),
+        )
+      } else if (this._host.isDefined && that._host.isDefined) {
+        List(
+          cmpOpt(this._userInfo, that._userInfo)(cmp),
+          cmpOpt(this._host, that._host)(_.compareToIgnoreCase(_)),
+          cmpOpt(this._port, that._port)(_ - _),  // change _port to js.UndefOr[Int] and fold into -1 only when getPort is called.
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: You actually need to make this change for this to compile :)

Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Oh, I missed.
Fixed in https://github.com/scala-js/scala-js/pull/3690/commits/dfe44551c417702d25aa841379360d442a559ca3

File: javalib/src/main/scala/java/net/URI.scala [hunk 7]
```diff
@@ -54,7 +54,7 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
   private val _authority = fld(AbsAuthority, RelAuthority).filter(_ != "")
   private val _userInfo = fld(AbsUserInfo, RelUserInfo)
   private val _host = fld(AbsHost, RelHost)
-  private val _port = fld(AbsPort, RelPort).fold(-1)(_.toInt)
+  private val _port = fld(AbsPort, RelPort)
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: `_port` should still be an `Int` field (add `.map(_.toInt)`)

File: javalib/src/main/scala/java/net/URI.scala [hunk 8]
```diff
@@ -101,41 +101,37 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
+      } else {
+        List(
+          cmpOpt(this._authority, that._authority)(cmp)
+        )
+      }
+    ) ++ List(
+      cmpOpt(this._path, that._path)(cmp),
+      cmpOpt(this._query, that._query)(cmp),
+      cmpOpt(this._fragment, that._fragment)(cmp)
+    )).find(_ != 0).getOrElse(0)
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: I'm unsure about the exact indentation rules of this. @sjrd will know :)

File: javalib/src/main/scala/java/net/URI.scala [hunk 9]
```diff
@@ -820,6 +820,14 @@ object URI {
     res
   }
 
+  /** Lower-cases host in `str` only if host exists. Used for hashing */
+  private def normalizeHost(str: String, host: js.UndefOr[String]): String = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: I'm not sure this is correct. What if the authority part is `FoO.CoM@FoO.CoM`?

File: javalib/src/main/scala/java/net/URI.scala (hunk 8 above)
Comment (sjrd) [reply javalib/src/main/scala/java/net/URI.scala]: It's mostly correct, except that every time something is inside `( ... )` (so in this case inside `List( ... )`, the indentation should be 4 spaces instead of 2. So it should be as follows:
```scala
    (List(
        cmpOpt(this._scheme, that._scheme)(_.compareToIgnoreCase(_)),
        this.isOpaque.compareTo(that.isOpaque) // A hierarchical URI is less than an opaque URI
    ) ++ (
        if (this.isOpaque()) {
          List(
              cmp(this._schemeSpecificPart, that._schemeSpecificPart)
          )
        } else if (this._host.isDefined && that._host.isDefined) {
          List(
              cmpOpt(this._userInfo, that._userInfo)(cmp),
              cmpOpt(this._host, that._host)(_.compareToIgnoreCase(_)),
              cmpOpt(this._port, that._port)(_.toInt - _.toInt)
          )
        } else {
          List(
              cmpOpt(this._authority, that._authority)(cmp)
          )
        }
    ) ++ List(
        cmpOpt(this._path, that._path)(cmp),
        cmpOpt(this._query, that._query)(cmp),
        cmpOpt(this._fragment, that._fragment)(cmp)
    )).find(_ != 0).getOrElse(0)
```

Comment (sjrd) [reply javalib/src/main/scala/java/net/URI.scala]: Btw you should use `:::` instead of `++` to concatenate `List`s. It is more efficient.

File: javalib/src/main/scala/java/net/URI.scala (hunk 7 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Addressed in https://github.com/scala-js/scala-js/pull/3690/commits/c0eaf768aa58dde27f84927961440e898479bf33

File: javalib/src/main/scala/java/net/URI.scala (hunk 8 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Indentation in https://github.com/scala-js/scala-js/pull/3690/commits/c2aaaa79c8380a5a1c8fdd21a6f6436ac16fb679
`:::` in https://github.com/scala-js/scala-js/pull/3690/commits/302c4b1d04b09b2e73be88f21e5ee1ef5e8d89c6

File: javalib/src/main/scala/java/net/URI.scala (hunk 9 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: That sounds right.
Addressed in https://github.com/scala-js/scala-js/pull/3690/commits/bba5ecc077c2298f087889cd701f6fe40339b537

**Comment by exoego** [comment]
I will squash recent minor commits when PR is good to go.

File: javalib/src/main/scala/java/net/URI.scala (hunk 9 above)
Comment (gzm0) [reply javalib/src/main/scala/java/net/URI.scala]: It feel this re-parses part of the URI again. Have you considered using the pre-parsed fields for this? (as in composing the hash code based on the present fields, rather than re-parsing and changing them)?

Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Done in https://github.com/scala-js/scala-js/pull/3690/commits/cfdb791567791ced8e75ee70a407a7eb3d0d30ae
I think *before* is simpler than *after* though re-parsing.

Comment (gzm0) [reply javalib/src/main/scala/java/net/URI.scala]: Personally I do prefer the new solution. @sjrd opinions?

Not sure we need the `List` with the `fold` rather than keeping mutable accumulation, but otherwise I like it.

Comment (sjrd) [reply javalib/src/main/scala/java/net/URI.scala]: Yes, I'm not quite sure what the old version was, but the new one is good.

I would also revert to using the mutable accumulator. It was more readable *and* more efficient.

File: javalib/src/main/scala/java/net/URI.scala [hunk 10]
```diff
@@ -166,10 +162,19 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
     var acc = URI.uriSeed
-    acc = mix(acc, _scheme.##) // scheme may not contain escapes
-    acc = mix(acc, normalizeEscapes(_schemeSpecificPart).##)
-    acc = mixLast(acc, _fragment.map(normalizeEscapes).##)
-
+    acc = mix(acc, _scheme.map(_.toLowerCase).##)
+    if (this.isOpaque()) {
+      acc = mix(acc, normalizeEscapes(this._schemeSpecificPart).##)
+    } else if (this._host.isDefined) {
+      acc = mix(acc, this._userInfo.##)
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: `userInfo` may contain escapes, so this needs to be normalized.

File: javalib/src/main/scala/java/net/URI.scala [hunk 11]
```diff
@@ -166,10 +162,19 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
+      acc = mix(acc, this._host.map(_.toLowerCase).##)
+      acc = mix(acc, this._port.##)
+    } else {
+      acc = mix(acc, this._authority.##)
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: `authority` may contain escapes, so this needs to be normalized.

File: javalib/src/main/scala/java/net/URI.scala [hunk 12]
```diff
@@ -166,10 +162,19 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
     import URI.normalizeEscapes
 
     var acc = URI.uriSeed
-    acc = mix(acc, _scheme.##) // scheme may not contain escapes
-    acc = mix(acc, normalizeEscapes(_schemeSpecificPart).##)
-    acc = mixLast(acc, _fragment.map(normalizeEscapes).##)
-
+    acc = mix(acc, _scheme.map(_.toLowerCase).##)
```
Comment (gzm0) [inline javalib/src/main/scala/java/net/URI.scala]: The comment about escapes should be retained.

File: javalib/src/main/scala/java/net/URI.scala [hunk 13]
```diff
@@ -101,41 +101,37 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
-      cmpOpt(this._path, that._path)
-    else if (this._query != that._query)
-      cmpOpt(this._query, that._query)
-    else
-      cmpOpt(this._fragment, that._fragment)
+    ) ::: List(
+        cmpOpt(this._path, that._path)(cmp),
+        cmpOpt(this._query, that._query)(cmp),
+        cmpOpt(this._fragment, that._fragment)(cmp)
+    )).find(_ != 0).getOrElse(0)
```
Comment (sjrd) [inline javalib/src/main/scala/java/net/URI.scala]: Looking back at this, I don't think the strategy of building a list with everything then filtering it is the best way. I'm all for functional code where it makes sense, but here it leads to completely unnecessary computations.

Have you considered a more imperative variant, like the following?
```scala
val cmpScheme = cmpOpt(this._scheme, that._scheme)(_.compareToIgnoreCase(_))
if (cmpScheme != 0)
  return cmpScheme
val cmpIsOpaque = this.isOpaque.compareTo(that.isOpaque) // A hierarchical URI is less than an opaque URI
if (cmpIsOpaque != 0)
  return cmpIsOpaque
if (this.isOpaque()) {
  ...
} else {
  ...
}
...
0
```
The new usage of `List` is also not in line with recent efforts to stop using Scala collections in the implementation of the javalib, which is what brought me back here, TBH.

**Comment by exoego** [comment]
> 
> 
> What's the status of this PR?

[The last CI error](https://scala-webapps.epfl.ch/sjsjenkins/blue/organizations/jenkins/scalajs-main/detail/PR-3690/13/pipeline#step-218-log-1) seems spurious, and I could not rerun so leavee it.

I will update per your request to avoid using Scala collections.

**Comment by exoego** [comment]
Updated and CI passes.

File: javalib/src/main/scala/java/net/URI.scala [hunk 14]
```diff
@@ -800,23 +827,23 @@ object URI {
...
-    var res = ""
-    while (i < str.length) {
-      if (str.charAt(i) == '%') {
-        assert(str.length > i + 2, "Invalid escape in URI")
-        res += str.substring(i, i+3).toUpperCase()
-        i += 3
-      } else {
-        res += str.substring(i, i+1)
-        i += 1
+  private def normalizeEscapes(maybeStr: js.UndefOr[String]): js.UndefOr[String] =
```
Comment (sjrd) [inline javalib/src/main/scala/java/net/URI.scala]: The body of this method should have `{}`.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala [hunk 15]
```diff
@@ -190,8 +193,33 @@ class URITest {
     assertTrue(x == x)
     assertTrue(y == y)
     assertTrue(z == z)
+  }
+
+  @Test def equals_and_hashCode_should_produces_same_result(): Unit = {
+    val equalsPairs: Seq[(URI,URI)] = Seq(
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: Missing space after `,`

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala [hunk 16]
```diff
@@ -190,8 +193,33 @@ class URITest {
...
+    val nonEqualPairs: Seq[(URI,URI)] = Seq(
+      (new URI("http://example.com/example-com"), new URI("http://Example.CoM/eXAMplE-cOm")),
+      (new URI("http://example.com@example.com"), new URI("http://EXAMPLE.COM@EXAMPLE.Com")),
+      (new URI("foo:helloWorld%6b%6C"), new URI("foo:helloWorld%6C%6b"))
+    )
+    nonEqualPairs.foreach { case (a, b) =>
+      assertNotEquals(a, b)
+      // Note: hashCode is not restricted to produce same result even if a.equals(b) is false.
+      assertNotEquals("a does not equal to b, but produces same hashCode. Pick different test data",
+        a.hashCode(), b.hashCode())
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: 4 spaces of indentation

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala [hunk 17]
```diff
@@ -337,10 +365,12 @@ class URITest {
   }
 
   @Test def should_provide_hashCode(): Unit = {
-    if (!executingInJVM) { // Fails on JDK6 and JDK7
-      assertEquals(new URI("http://example.com/asdf%6a").hashCode,
-          new URI("http://example.com/asdf%6A").hashCode)
-    }
+    assertEquals(new URI("http://example.com/asdf%6a").hashCode,
+      new URI("http://example.com/asdf%6A").hashCode)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: 4 spaces of indentation (same for the two below)

Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: Fixed

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala (hunk 16 above)
Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: Fixed

File: javalib/src/main/scala/java/net/URI.scala (hunk 14 above)
Comment (exoego) [reply javalib/src/main/scala/java/net/URI.scala]: Fixed

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala (hunk 15 above)
Comment (exoego) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala]: Fixed
